minidump-processor/src/op_analysis.rs

```diff
@@ -0,0 +1,733 @@
+///     to indicate that end-of-file is immediately hit
+///
+/// Implementing this marker trait for a type that doesn't meet this requirements has safe but
+/// unspecified behavior
+///
+/// As these requirements are fairly lenient, this means that any `File` or `&[u8]` that contains
+/// a flat memory dump of process memory may implement this trait
+///
+pub trait SparseMemoryStream: Read + Seek {}
+
+/// Read the instruction bytes that were being run by the given thread
+///
+/// Use the given `context` to attempt to read `1 <= n <= MAX_INSTRUCTION_LENGTH`
+/// bytes at the instruction pointer from the byte stream passed in `memory`, which is any type
+/// that implements `Seek + Read` and is meant to represent the memory of the target machine
+///
+///
+///
+/// # Errors
+///
+/// This may fail if the underlying byte stream fails to `seek()` or `read()`
+///
+pub fn get_thread_instruction_bytes(
+    context: &MinidumpContext,
+    memory: &mut impl SparseMemoryStream,
+) -> Result<Vec<u8>, OpAnalysisError> {
+    let instruction_pointer = context.get_instruction_pointer();
+
+    memory
+        .seek(SeekFrom::Start(instruction_pointer))
+        .map_err(OpAnalysisError::ReadThreadInstructionFailed)?;
+
+    // We use MAX_INSTRUCTION_LENGTH here as an optimization to avoid allocating and copying
+    // more bytes than needed to produce an instruction
+
+    let mut buffer = Vec::with_capacity(MAX_INSTRUCTION_LENGTH.into());
+
+    memory
+        .take(MAX_INSTRUCTION_LENGTH.into())
+        .read_to_end(&mut buffer)
+        .map_err(OpAnalysisError::ReadThreadInstructionFailed)?;
+
+    Ok(buffer)
+}
+
+/// Details about a memory access performed by an instruction
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct MemoryAccess {
+    /// The address of the memory access
+    pub address: u64,
+    /// The size of the memory access
+    ///
+    /// Note that this is optional, as there are weird instructions that do not know the size
+    /// of their memory accesses without more complex context
+    pub size: Option<u8>,
+}
+
+/// Determine the memory accesses that would occur with the given instruction and context
+///
+/// Architectures like x86 allow for complex arithmetic involving multiple registers to be
+/// used to determine a target memory address, and allow for single operations involving
+/// multiple memory locations (bit blitting)
+///
+/// If the given `instruction_bytes` contain a valid instruction, and the given `context` contains
+/// valid values for all the registers used for address calculation, this function will return a
+/// (possibly-empty) list of all the memory accesses for the instruction
+///
+/// ⚠ NOTE ⚠ - Certain instructions like PUSH are not considered to access memory, even though
+/// they do technically write to the stack. Generally this shouldn't be of much concern, but
+/// there may be some situations where this detail matters
```

Comment:
hi! i'm really curious about this comment in particular -- there was [a bug](https://github.com/iximeow/yaxpeax-x86/blob/no-gods-no-/CHANGELOG#L83-L88) i'd fixed in yaxpeax-x86 where _implicit_ memory accesses (`push rax`) were not reported, but _explicit_ memory accesses (`push qword [rax]`) were. i added [a few test cases](https://github.com/iximeow/yaxpeax-x86/blob/no-gods-no-/test/long_mode/operand.rs#L32-L47) when fixing that issue, so now yaxpeax-x86 should be consistently reporting an instruction accesses memory even if it's not through an explicit operand.

since this comment seems similar to a bug i'd fixed in `1.1`, and you're using `1.1`, and i don't see explicit mention of instructions like `push` in this PR, i'm curious if you'd initially used an older version, found the bug, then added the comment referencing the now-fixed yaxpeax behavior?

(ps: it's very cool to see yax is useful here! use in tools is exactly what i'd had in mind for the crates :tada:)

Reply:
ah nvm! i see the logic now: there won't be a memory-accessing operand to iterate over, so you won't record a memory access. and if you _did_ want to record a memory access you'd need to consult the `Opcode` to know what registers to use in an address calculation (like.. `rsp`).

so then there's an addendum to offer to this comment if you're interested: cases like `push qword [rax]` _would_ be considered as accessing memory, but only to read the operand, not for the implied push afterwards :grin:

Reply:
(just for historical record, ixi and i discussed this and it seems like the set of instructions with implicit memory operands are pretty rare and enumerable. I believe it's just the various flavours of push, pop, call, save, restore. I'm perfectly fine with landing this as is though, and waiting for yaxpeax to provide some general API for enumerating implicit operands.)

Reply:
Yeah -- I looked at [the list of amd64 opcodes in iximeow's library](https://docs.rs/yaxpeax-x86/latest/yaxpeax_x86/long_mode/enum.Opcode.html), had a minor heart attack, and decided that perhaps some things were better left to experts 😂